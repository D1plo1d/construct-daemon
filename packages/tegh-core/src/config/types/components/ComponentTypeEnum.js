import { List } from 'immutable'

export const CONTROLLER = 'CONTROLLER'
export const TOOLHEAD = 'TOOLHEAD'
export const BUILD_PLATFORM = 'BUILD_PLATFORM'
export const FAN = 'FAN'

const ComponentTypeEnum = List([
  CONTROLLER,
  TOOLHEAD,
  BUILD_PLATFORM,
  FAN,
])

export default ComponentTypeEnum
