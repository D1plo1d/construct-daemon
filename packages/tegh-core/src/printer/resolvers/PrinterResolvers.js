import isIdle from '../../spool/selectors/isIdle'
import getComponentsState from '../selectors/getComponentsState'
import getPluginModels from '../../config/selectors/getPluginModels'

const PrinterResolvers = {
  Printer: {
    id: source => source.config.printer.id,
    name: source => getPluginModels(source.config).getIn(['tegh-core', 'name']),
    heaters: source => getComponentsState(source).heaters.toList(),
    targetTemperaturesCountdown: source => getComponentsState(source).targetTemperaturesCountdown,
    activeExtruderID: source => getComponentsState(source).activeExtruderID,
    fans: source => getComponentsState(source).fans.toList(),
    status: (source) => {
      if (!isIdle(source.spool)) return 'PRINTING'
      const { status } = source.status
      return status.substring(status.lastIndexOf('/') + 1)
    },
    error: source => source.status.error,
    macroDefinitions: source => source.macros.macros.keys(),
    logEntries: (source, args) => {
      let entries = source.log.get('logEntries')
      if (args.level != null) {
        entries = entries.filter(log => log.level === args.level)
      }
      if (args.source != null) {
        entries = entries.filter(log => log.source === args.source)
      }
      return entries.toArray()
    },
  },
}

export default PrinterResolvers
