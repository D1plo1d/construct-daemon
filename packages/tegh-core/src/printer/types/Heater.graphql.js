import tql from 'typiql'
import snl from 'strip-newlines'
import {
  GraphQLObjectType,
} from 'graphql'

import getHeaterConfigs from '../../config/selectors/getHeaterConfigs'

const HeaterType = new GraphQLObjectType({
  name: 'Heater',
  description: 'A heated bed or extruder',
  fields: () => ({
    id: {
      type: tql`ID!`,
      resolve: source => `${source.id}DynamicData`,
    },
    name: {
      type: tql`String!`,
      resolve: (source, args, { store }) => {
        const state = store.getState()
        return getHeaterConfigs(state.config).get(source.id).name
      },
    },
    type: {
      type: tql`String!`,
      resolve: (source, args, { store }) => {
        const state = store.getState()
        return getHeaterConfigs(state.config).get(source.id).type
      },
    },
    targetTemperature: {
      type: tql`Float`,
      description: snl`
        The target temperature in °C for this heater. The heater will
        attempt to make the current_temp equal to this temperature.
      `,
    },
    currentTemperature: {
      type: tql`Float!`,
      description: snl`
        The current temperature in °C recorded by the heater’s thermocouple
        or thermister.
      `,
    },
    blocking: {
      type: tql`Boolean!`,
      description: snl`
        True if the printer is waiting on this heater to reach it’s
        targetTemp and preventing any more gcodes from executing until it does.
      `,
    },
  }),
})

export default HeaterType
