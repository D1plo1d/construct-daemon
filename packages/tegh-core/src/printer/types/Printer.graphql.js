import { GraphQLObjectType } from 'graphql'
import tql from 'typiql'
import snl from 'strip-newlines'

import isIdle from '../../spool/selectors/isIdle'
import getDriverState from '../selectors/getDriverState'

import PrinterStatusEnum from './PrinterStatusEnum.graphql'
import HeaterType from './Heater.graphql'
import FanType from './Fan.graphql'
import LogEntryType from '../../log/types/LogEntry.graphql'
import MacroDefinitionType from '../../config/types/MacroDefinition.graphql'
import PrinterErrorType from './PrinterError.graphql'

const PrinterGraphQL = new GraphQLObjectType({
  name: 'Printer',
  fields: () => ({
    id: {
      type: tql`ID!`,
      resolve(source) {
        return source.config.id
      },
    },
    name: {
      type: tql`String!`,
      resolve(source) {
        return source.config.name
      },
    },
    heaters: {
      type: tql`[${HeaterType}!]!`,
      resolve(source) {
        return getDriverState(source).peripherals.heaters.values()
      },
    },
    targetTemperaturesCountdown: {
      type: tql`Float`,
      description: snl`
        The estimated number of seconds until the heater(s) reach their
        targetTemperature.
      `,
      resolve(source) {
        return getDriverState(source).peripherals.targetTemperaturesCountdown
      },
    },
    activeExtruderID: {
      type: tql`String`,
      description: snl`
        The active extruder ID
      `,
      resolve(source) {
        return getDriverState(source).peripherals.activeExtruderID
      },
    },
    fans: {
      type: tql`[${FanType}!]!`,
      resolve(source) {
        return getDriverState(source).peripherals.fans.values()
      },
    },
    status: {
      type: tql`${PrinterStatusEnum}!`,
      resolve: (source) => {
        if (!isIdle(source.spool)) return 'PRINTING'
        const { status } = source.status
        return status.substring(status.lastIndexOf('/') + 1)
      },
    },
    error: {
      type: tql`${PrinterErrorType}`,
      resolve: source => source.status.error,
    },
    macroDefinitions: {
      type: tql`[${MacroDefinitionType}!]!`,
      resolve: source => source.macros.keys(),
    },
    logEntries: {
      type: tql`[${LogEntryType}!]`,
      args: {
        level: {
          type: tql`String`,
        },
        source: {
          type: tql`String`,
        },
      },
      resolve(source, args) {
        let entries = source.log.get('entries')
        if (args.level != null) {
          entries = entries.filter(log => log.level === args.level)
        }
        if (args.source != null) {
          entries = entries.filter(log => log.source === args.source)
        }
        return entries.toArray()
      },
    },
  }),
})

export default PrinterGraphQL
