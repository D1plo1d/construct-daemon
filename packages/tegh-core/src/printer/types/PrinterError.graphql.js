import {
  GraphQLObjectType,
} from 'graphql'
import tql from 'typiql'
import snl from 'strip-newlines'

const PrinterErrorType = new GraphQLObjectType({
  name: 'PrinterError',
  fields: () => ({
    code: {
      type: tql`String!`,
      description: snl`
        A machine-readable code indicating the type of error
      `,
    },
    message: {
      type: tql`String!`,
      description: snl`
        A human-readable description of the error
      `,
    },
  }),
})

export default PrinterErrorType
