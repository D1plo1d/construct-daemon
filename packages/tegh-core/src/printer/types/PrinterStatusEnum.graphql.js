import snl from 'strip-newlines'
import {
  GraphQLEnumType,
} from 'graphql'

const PrinterStatusEnum = new GraphQLEnumType({
  name: 'PrinterStatusEnum',
  values: {
    CONNECTING: {
      description: snl`
        The printer is being initialized.
        Attempting to spool anything will result in an error.
      `,
    },
    READY: {
      description: snl`
        The printer is connected and able to spool tasks/jobs.
      `,
    },
    PRINTING: {
      description: snl`
        The printer is printing a job.
        Attempting to spool anything will result in an error.
      `,
    },
    DISCONNECTED: {
      description: snl`
        The printer is disconnected or turned off.
        Attempting to spool anything will result in an error.
      `,
    },
    ERRORED: {
      description: snl`
        The printer is being initialized.
        Attempting to spool anything except for an
        emergency macro (ie. reset) will result in an error. Spool the \`reset\`
        macro to reset the error and change the status to \`CONNECTING\`.
      `,
    },
    ESTOPPED: {
      description: snl`
        The printer is estopped. Attempting to spool anything except for an
        emergency macro (ie. reset) will result in an error. Spool the \`reset\`
        macro to reset the estop and change the status to \`CONNECTING\`.
      `,
    },
  },
})

export default PrinterStatusEnum
