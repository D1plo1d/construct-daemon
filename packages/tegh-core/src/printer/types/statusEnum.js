export const ERRORED = 'tegh/printer/status/ERRORED'
export const ESTOPPED = 'tegh/printer/status/ESTOPPED'
export const DISCONNECTED = 'tegh/printer/status/DISCONNECTED'
export const CONNECTING = 'tegh/printer/status/CONNECTING'
export const READY = 'tegh/printer/status/READY'
