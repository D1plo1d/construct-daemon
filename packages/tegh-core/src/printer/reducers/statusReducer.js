import { Record } from 'immutable'

import { SPOOL_TASK } from '../../spool/actions/spoolTask'
import { ESTOP } from '../actions/estop'
import { DRIVER_ERROR } from '../actions/driverError'
import { PRINTER_READY } from '../actions/printerReady'

import { EMERGENCY } from '../../spool/types/PriorityEnum'

import { PRINTER_CONNECTING } from '../actions/printerConnecting'
import { PRINTER_DISCONNECTED } from '../actions/printerDisconnected'

import {
  ERRORED,
  ESTOPPED,
  DISCONNECTED,
  CONNECTING,
  READY,
} from '../types/statusEnum'

const initialState = Record({
  status: DISCONNECTED,
  error: null,
})()

const statusReducer = (state = initialState, action) => {
  switch (action.type) {
    case DRIVER_ERROR: {
      return initialState
        .set('status', ERRORED)
        .set('error', action.payload)
    }
    case ESTOP: {
      return initialState.set('status', ESTOPPED)
    }
    case PRINTER_DISCONNECTED: {
      return initialState.set('status', DISCONNECTED)
    }
    case PRINTER_CONNECTING: {
      return initialState.set('status', CONNECTING)
    }
    case PRINTER_READY: {
      return initialState.set('status', READY)
    }
    case SPOOL_TASK: {
      if (
        state.status !== READY
        && action.payload.task.priority !== EMERGENCY
      ) {
        const err = (
          'Only emergency tasks can be spooled when the machine is not ready.'
        )
        throw new Error(err)
      }
      return state
    }
    default: {
      return state
    }
  }
}

export default statusReducer
