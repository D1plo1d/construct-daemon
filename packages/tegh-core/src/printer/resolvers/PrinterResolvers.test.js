import { Map, Record } from 'immutable'
import typeDefs from 'tegh-schema'

import deterministicTestSetup from '../../util/testing/deterministicTestSetup'
import snapshotTestResolvers from '../../util/testing/snapshotTestResolvers'

import { CONNECTING } from '../types/statusEnum'
import { INFO } from '../../log/types/logLevelEnum'
import { TOOLHEAD, FAN } from '../../config/types/components/ComponentTypeEnum'

import { initialState as spoolInitialState } from '../../spool/reducers/spoolReducer'
import { initialState as macrosInitialState } from '../../macros/reducers/macrosReducer'
import { initialState as logInitialState } from '../../log/reducers/logReducer'
import { initialState as statusInitialState } from '../reducers/statusReducer'
import { MockConfig } from '../../config/types/Config'

import PrinterResolvers from './PrinterResolvers'

describe('PrinterResolvers', () => {
  it('presents a consistent API', () => {
    deterministicTestSetup()

    const state = Record({
      config: MockConfig(),
      spool: spoolInitialState,
      status: statusInitialState.set('status', CONNECTING),
      macros: macrosInitialState.set('macros', Map({ myMacro: () => null })),
      log: logInitialState.setIn(['logEntries', 0], {
        source: 'SERIAL',
        level: INFO,
        message: 'This is an example log',
      }),
      'tegh-driver-serial-gcode': {
        components: Record({
          targetTemperaturesCountdown: 3,
          activeExtruderID: 'heater1ID',
          byID: Map({
            heater1Address: Record({
              id: 'heater1ID',
              type: TOOLHEAD,
            })(),
            fan1Address: Record({
              id: 'fan1ID',
              type: FAN,
            })(),
          }),
        })(),
      },
    })()

    snapshotTestResolvers({
      typeDefs,
      typeName: 'Printer',
      resolvers: PrinterResolvers,
      rootValue: state,
      contextValue: {
        store: {
          getState: () => state,
        },
      },
    })
  })
})
