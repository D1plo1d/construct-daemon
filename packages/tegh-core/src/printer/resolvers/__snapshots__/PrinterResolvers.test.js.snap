// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobResolvers presents a consistent API 1`] = `
Object {
  "_testPrinter": Object {
    "activeExtruderID": "heater1ID",
    "error": null,
    "fans": Array [
      Object {
        "id": "fan1ID",
      },
    ],
    "heaters": Array [
      Object {
        "id": "heater1ID",
      },
    ],
    "id": "aaxcvxcvcxv-bvb-csdf234231",
    "logEntries": Array [
      Object {},
    ],
    "macroDefinitions": Array [
      Object {},
    ],
    "name": "Left Home Lulzbot",
    "status": "CONNECTING",
    "targetTemperaturesCountdown": 3,
  },
}
`;
